searches/binary_search.nim

```diff
@@ -0,0 +1,99 @@
+## Binary Search
+## =============
+## Binary search is an efficient searching algorithm. It is faster than linear 
+## search except for small arrays. However, binary search can only be used on sorted arrays.
+## If the array is not sorted, it must be sorted first before binary search can be applied.
+##
+## Binary search starts by comparing the key value with the middle element of the array.
+## If the key value matches the middle element, the search is complete.
+## If the key value is less than the middle element, the search continues on the lower half of the array.
+## If the key value is greater than the middle element, the search continues on the upper half of the array.
+## This process repeats until the middle element matches the key value or the search space is exhausted.
```

Comment:
Add a *small* note on other approaches like linear search and search in a tree. See: https://en.wikipedia.org/wiki/Binary_search_algorithm#Binary_search_versus_other_schemes